### ISSUE TITLE
[java] [regression from 6.55.0] UnusedAssignment false-positive

### ISSUE THREAD

**Issue by kohlschuetter** [pmd/pmd#4617]
**Affects PMD Version:** 

**Rule:**

[https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#unusedassignment](UnusedAssignment)

**Description:**

**Code Sample demonstrating the issue:**

```java
SocketServer.java
```

**Expected outcome:**

PMD reports a violation, but that's wrong. That's a false positive.
```
SocketServer.java:446:	UnusedAssignment:	The value assigned to field 'timeoutFuture' is never used (overwritten on line 449)
```
"timeoutFuture" is set to null in line 446 because another thread may be accessing it. While there is a potential short time window between line 445 and 446, there can be a significantly longer time interval between calling TIMEOUTS.schedule and assigning its return value to timeoutFuture in line 449).

We can check timeoutFuture from several threads and multiple methods (`stop()` and `startThenStopAfter()`), so this shouldn't be an error (and it wasn't reported by 6.55.0).

**Running PMD through:** CLI

file and ruleset attached
[file-and-ruleset2.zip](https://github.com/pmd/pmd/files/11909565/file-and-ruleset2.zip)

**Comment by oowekyala** [comment]
As a preface, this rule is not smart enough to understand threads or timing, and given thread sensitive code, it may give FPs, hopefully rarely. This wasn't reported in PMD 6.55.0 because at that time the rule didn't check assignments to fields at all. Support for checking fields is still rough on the edges (like you found out in #4618).

In this instance though, all accesses to the `timeoutFuture` field are guarded by `synchronized (this)` so the field cannot be accessed concurrently by different threads, can it? Unless I'm mistaken, you could avoid nulling out the field on line 446, as the rule suggests.

**Comment by kohlschuetter** [comment]
@oowekyala Right, the code is currently guarded with `synchronized(this)`, so it should be safe. However, the warning also occurs when `synchronized` is removed.

It's not only a threading problem because it theoretically also occur from within the same thread if the method called in `return (this.timeoutFuture =` accessed `this.timeoutFuture` itself (at that point, the member variable is still `null`).

It would be helpful to mention threading/timing in the documentation for `UnusedAssignment` as potential source for a false-positive; it looks like the documentation hasn't been updated for PMD 7 yet.

**Comment by oowekyala** [comment]
> It's not only a threading problem because it theoretically also occur from within the same thread if the method called in return (this.timeoutFuture = accessed this.timeoutFuture itself (at that point, the member variable is still null).

The rule tries to address this already. Every method call that somehow has access to `this` implicitly marks all fields of `this` as used (`this` escapes). In this case `TIMEOUTS.schedule(..)` is not such a method call. However, constructing an anonymous class like `new Callable` implicitly captures `this`, which is currently not considered by the rule as an escape. We could fix your problem by extending the definition of escape to that. But we should keep in mind the rule's memory model is not rock solid, and FPs may still occur. I've documented that in [b92e751](https://github.com/oowekyala/pmd/commit/b92e7512eea112458f048c0c929284f3a51465ec)
